File: src/TokenDock/UsageApiClient.cs
using System.Globalization;
using System.Net;
using System.Net.Http.Headers;
using System.Text.Json;

namespace TokenDock;

/// <summary>OpenCode Go 用量接口客户端。接口：GET https://opencode.ai/zen/go/v1/usage（Bearer 鉴权）。</summary>
public sealed class UsageApiClient : IDisposable
{
    public const string DefaultEndpoint = "https://opencode.ai/zen/go/v1/usage";

    private readonly HttpClient _http;

    public UsageApiClient() : this(CreateHttpClient())
    {
    }

    public UsageApiClient(HttpClient http) => _http = http;

    public void Dispose() => _http.Dispose();

    private static HttpClient CreateHttpClient()
    {
        var http = new HttpClient(new SocketsHttpHandler
        {
            PooledConnectionLifetime = TimeSpan.FromMinutes(10),
            ConnectTimeout = TimeSpan.FromSeconds(10),
        })
        {
            Timeout = TimeSpan.FromSeconds(15),
        };
        http.DefaultRequestHeaders.UserAgent.ParseAdd("TokenDock/1.0");
        http.DefaultRequestHeaders.Accept.ParseAdd("application/json");
        return http;
    }

    /// <summary>使用 Bearer 密钥拉取用量。任何网络异常都会转为失败结果，不会抛出（程序主动取消除外）。</summary>
    public async Task<FetchResult> FetchAsync(string? apiKey, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(apiKey))
            return FetchResult.Fail(FetchFailureKind.NotConfigured,
                "尚未设置 API 密钥，请右键托盘图标选择「设置密钥」。");

        using var request = new HttpRequestMessage(HttpMethod.Get, DefaultEndpoint);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", apiKey);

        HttpResponseMessage response;
        try
        {
            response = await _http.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
        }
        catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
        {
            return FetchResult.Fail(FetchFailureKind.NetworkError, "连接超时（15 秒内未收到响应），请检查网络。");
        }
        catch (OperationCanceledException)
        {
            throw;
        }
        catch (HttpRequestException ex)
        {
            return FetchResult.Fail(FetchFailureKind.NetworkError, $"网络连接失败：{ex.Message}");
        }
        catch (Exception ex)
        {
            return FetchResult.Fail(FetchFailureKind.NetworkError, $"请求出现异常：{ex.GetType().Name} {ex.Message}");
        }

        using (response)
        {
            string body;
            try
            {
                body = await response.Content.ReadAsStringAsync(cancellationToken);
            }
            catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
            {
                return FetchResult.Fail(FetchFailureKind.NetworkError, "读取响应超时，请检查网络。");
            }

            return Classify(response.StatusCode, body);
        }
    }

    /// <summary>按状态码归类接口结果（公开静态以便测试）。</summary>
    public static FetchResult Classify(HttpStatusCode statusCode, string body)
    {
        var detail = TryExtractServerMessage(body);
        return statusCode switch
        {
            HttpStatusCode.OK => ParseSuccessBody(body),
            HttpStatusCode.Unauthorized => FetchResult.Fail(FetchFailureKind.InvalidKey,
                "API 密钥无效或已失效（HTTP 401），请重新设置密钥。", detail),
            HttpStatusCode.Forbidden => FetchResult.Fail(FetchFailureKind.NoSubscription,
                "无权访问该接口（HTTP 403），可能尚未订阅 OpenCode Go 或权限不足。", detail),
            HttpStatusCode.PaymentRequired => FetchResult.Fail(FetchFailureKind.NoSubscription,
                "订阅不可用（HTTP 402），请确认 OpenCode Go 订阅状态。", detail),
            HttpStatusCode.TooManyRequests => FetchResult.Fail(FetchFailureKind.RateLimited,
                "请求过于频繁，已被限流（HTTP 429），请稍后再试。", detail),
            var code when (int)code >= 500 => FetchResult.Fail(FetchFailureKind.ServerError,
                $"OpenCode 服务端异常（HTTP {(int)code}），请稍后重试。", detail),
            var code => FetchResult.Fail(FetchFailureKind.ServerError,
                $"接口返回异常状态码（HTTP {(int)code}）。", detail),
        };
    }

    /// <summary>
    /// 解析 HTTP 200 的响应体（公开静态以便测试）。仅读取真实返回的 usage.rolling / weekly / monthly
    /// 中的 percent、status、resetsAt 字段；剩余百分比 = 100 - 已用百分比。不推断任何 Token 或请求次数额度。
    /// </summary>
    public static FetchResult ParseSuccessBody(string body)
    {
        JsonDocument doc;
        try
        {
            doc = JsonDocument.Parse(string.IsNullOrWhiteSpace(body) ? "null" : body);
        }
        catch (JsonException)
        {
            return FetchResult.Fail(FetchFailureKind.ParseError, "接口返回内容不是有效的 JSON。", Snip(body));
        }

        using (doc)
        {
            var root = doc.RootElement;
            if (root.ValueKind != JsonValueKind.Object
                || !root.TryGetProperty("usage", out var usage)
                || usage.ValueKind != JsonValueKind.Object)
            {
                return FetchResult.Fail(FetchFailureKind.ParseError,
                    "接口返回格式不符合预期（缺少 usage 字段）。", Snip(body));
            }

            var rolling = ParseWindow(usage, "rolling");
            var weekly = ParseWindow(usage, "weekly");
            var monthly = ParseWindow(usage, "monthly");
            if (rolling is null && weekly is null && monthly is null)
            {
                return FetchResult.Fail(FetchFailureKind.ParseError,
                    "接口返回中缺少 rolling/weekly/monthly 用量数据，可能当前无订阅，或接口格式已变更。", Snip(body));
            }

            return FetchResult.Ok(new UsageData { Rolling = rolling, Weekly = weekly, Monthly = monthly });
        }
    }

    private static UsageWindow? ParseWindow(JsonElement usage, string name)
    {
        if (!usage.TryGetProperty(name, out var w) || w.ValueKind != JsonValueKind.Object)
            return null;

        double percent = double.NaN;
        string? status = null;
        DateTimeOffset? resetsAt = null;

        if (w.TryGetProperty("percent", out var p))
        {
            if (p.ValueKind == JsonValueKind.Number && p.TryGetDouble(out var d))
                percent = d;
            else if (p.ValueKind == JsonValueKind.String
                     && double.TryParse(p.GetString(), NumberStyles.Float, CultureInfo.InvariantCulture, out var ds))
                percent = ds;
        }

        if (w.TryGetProperty("status", out var s) && s.ValueKind == JsonValueKind.String)
            status = s.GetString();

        if (w.TryGetProperty("resetsAt", out var r))
        {
            if (r.ValueKind == JsonValueKind.String
                && DateTimeOffset.TryParse(r.GetString(), CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal, out var dto))
                resetsAt = dto;
            else if (r.ValueKind == JsonValueKind.Number && r.TryGetInt64(out var unix))
                resetsAt = DateTimeOffset.FromUnixTimeSeconds(unix);
        }

        return new UsageWindow { PercentUsed = percent, Status = status, ResetsAt = resetsAt };
    }

    /// <summary>从错误响应中提取 error.message。实测错误体形如 {"type":"error","error":{"type":"AuthError","message":"..."}}。</summary>
    private static string? TryExtractServerMessage(string body)
    {
        try
        {
            using var doc = JsonDocument.Parse(body);
            if (doc.RootElement.ValueKind == JsonValueKind.Object
                && doc.RootElement.TryGetProperty("error", out var err)
                && err.ValueKind == JsonValueKind.Object
                && err.TryGetProperty("message", out var msg)
                && msg.ValueKind == JsonValueKind.String)
            {
                return msg.GetString();
            }
        }
        catch (JsonException)
        {
            // 非 JSON 错误体，忽略
        }

        return null;
    }

    private static string? Snip(string body)
        => string.IsNullOrWhiteSpace(body) ? null
            : body.Length <= 200 ? body
            : body[..200] + "…";
}

File: android/app/src/main/java/com/tokendock/app/data/UsageApiClient.kt
package com.tokendock.app.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL

/**
 * OpenCode Go 官方额度客户端：GET https://opencode.ai/zen/go/v1/usage（Bearer 鉴权）。
 *
 * 仅读取真实返回的 usage.rolling / weekly / monthly 中的 percent、status、resetsAt；
 * 不推断任何 Token 数或额度数字——接口没返回的一律为 null（界面显示「未知」）。
 */
class UsageApiClient(private val endpoint: String = DEFAULT_ENDPOINT) {

    suspend fun fetch(apiKey: String?): FetchResult = withContext(Dispatchers.IO) {
        if (apiKey.isNullOrBlank()) {
            return@withContext FetchResult.Fail(
                FetchFailureKind.NotConfigured,
                "尚未设置 API 密钥，请在设置中填写。",
            )
        }

        val connection = try {
            (URL(endpoint).openConnection() as HttpURLConnection).apply {
                requestMethod = "GET"
                connectTimeout = 10_000
                readTimeout = 15_000
                setRequestProperty("Authorization", "Bearer $apiKey")
                setRequestProperty("Accept", "application/json")
                setRequestProperty("User-Agent", "TokenDock-Android/1.0")
            }
        } catch (e: IOException) {
            return@withContext FetchResult.Fail(FetchFailureKind.NetworkError, "网络连接失败：${e.message}")
        } catch (e: Exception) {
            return@withContext FetchResult.Fail(FetchFailureKind.NetworkError, "请求出现异常：${e.javaClass.simpleName}")
        }

        try {
            val code = connection.responseCode
            val body = (if (code in 200..299) connection.inputStream else connection.errorStream)
                ?.bufferedReader()?.use { it.readText() } ?: ""
            classify(code, body)
        } catch (e: IOException) {
            FetchResult.Fail(FetchFailureKind.NetworkError, "读取响应失败：${e.message}")
        } finally {
            connection.disconnect()
        }
    }

    companion object {
        const val DEFAULT_ENDPOINT = "https://opencode.ai/zen/go/v1/usage"

        /** 按状态码归类（与桌面端同一套语义）。 */
        fun classify(statusCode: Int, body: String): FetchResult {
            val detail = extractServerMessage(body)
            return when (statusCode) {
                200 -> parseSuccessBody(body)
                401 -> FetchResult.Fail(FetchFailureKind.InvalidKey, "API 密钥无效或已失效（HTTP 401），请重新设置。", detail)
                402 -> FetchResult.Fail(FetchFailureKind.NoSubscription, "订阅不可用（HTTP 402），请确认 OpenCode Go 订阅状态。", detail)
                403 -> FetchResult.Fail(FetchFailureKind.NoSubscription, "无权访问该接口（HTTP 403），可能尚未订阅或权限不足。", detail)
                429 -> FetchResult.Fail(FetchFailureKind.RateLimited, "请求过于频繁（HTTP 429），请稍后再试。", detail)
                else -> if (statusCode >= 500) {
                    FetchResult.Fail(FetchFailureKind.ServerError, "服务端异常（HTTP $statusCode），请稍后重试。", detail)
                } else {
                    FetchResult.Fail(FetchFailureKind.ServerError, "接口返回异常状态码（HTTP $statusCode）。", detail)
                }
            }
        }

        /** 解析 HTTP 200 响应体；格式不符时返回 ParseError（不编造数据）。 */
        fun parseSuccessBody(body: String): FetchResult {
            val root = try {
                JSONObject(body.ifBlank { "null" })
            } catch (e: JSONException) {
                return FetchResult.Fail(FetchFailureKind.ParseError, "接口返回内容不是有效的 JSON。", snip(body))
            }

            val usage = root.optJSONObject("usage")
                ?: return FetchResult.Fail(
                    FetchFailureKind.ParseError,
                    "接口返回格式不符合预期（缺少 usage 字段）。",
                    snip(body),
                )

            val rolling = parseWindow(usage, "rolling")
            val weekly = parseWindow(usage, "weekly")
            val monthly = parseWindow(usage, "monthly")
            if (rolling == null && weekly == null && monthly == null) {
                return FetchResult.Fail(
                    FetchFailureKind.ParseError,
                    "接口返回中缺少 rolling/weekly/monthly 用量数据，可能当前无订阅，或接口格式已变更。",
                    snip(body),
                )
            }

            return FetchResult.Ok(
                UsageData(
                    rolling = rolling,
                    weekly = weekly,
                    monthly = monthly,
                    fetchedAtMillis = System.currentTimeMillis(),
                ),
            )
        }

        private fun parseWindow(usage: JSONObject, name: String): UsageWindow? {
            val window = usage.optJSONObject(name) ?: return null
            val percent = when {
                window.has("percent") && !window.isNull("percent") -> {
                    val raw = window.opt("percent")
                    when (raw) {
                        is Number -> raw.toDouble()
                        is String -> raw.toDoubleOrNull()
                        else -> null
                    }
                }
                else -> null
            }
            val status = window.optString("status", "").ifBlank { null }
            val resetsAt = window.opt("resetsAt")?.let { value ->
                when (value) {
                    is String -> parseIsoMillis(value)
                    is Number -> value.toLong()
                    else -> null
                }
            }
            return UsageWindow(percentUsed = percent, status = status, resetsAtMillis = resetsAt)
        }

        /** 解析 ISO8601（形如 2026-09-17T12:49:17.815Z）；minSdk 28 起 java.time 可直接使用。 */
        fun parseIsoMillis(text: String): Long? = try {
            java.time.Instant.parse(text).toEpochMilli()
        } catch (e: Exception) {
            null
        }

        /** 从错误体提取 error.message（实测形态 {"type":"error","error":{"type":"AuthError","message":"..."}}）。 */
        fun extractServerMessage(body: String): String? = try {
            JSONObject(body).optJSONObject("error")?.optString("message")?.ifBlank { null }
        } catch (e: JSONException) {
            null
        }

        private fun snip(body: String): String? = when {
            body.isBlank() -> null
            body.length <= 200 -> body
            else -> body.take(200) + "…"
        }
    }
}
